mountinfo/mounted_linux.go

```diff
@@ -7,6 +7,82 @@ import (
 	"golang.org/x/sys/unix"
 )
 
+// MountedBind is a Linux-specific function, similar to Mounted, but with
+// additional workarounds to decrease the possibility of misdetecting
+// a bind mount.
+//
+// On Linux (before kernel 5.8) bind mount detection is not reliable (false
+// negatives are possible) on systems that perform lots of mounts and unmounts,
+// due to a kernel issue. This function tries to mitigate that issue by
+// re-reading mountinfo, decreasing the possibility of a false negative.
+//
+// This function increases system load. It should only be used for bind
+// mounts, if a user wants to sacrifice system performance for higher reliability.
+func MountedBind(path string) (bool, error) {
+	return mountedWithFallback(path, mountedByMountinfoWithRetries)
+}
+
+// mountedByMountinfoWithRetries is like mountedByMountinfo, with
+// the addition of a workaround for a kernel bug (fixed in Linux 5.8).
+// This kernel bug is essentially a race between reading mountinfo
+// and performing an unmount. In the case next to be read mount entry is
+// removed from the kernel data structures, it prematurely returns EOF.
+// As a result, sometimes only a half of mountinfo is read, leading to
+// inability to find a legitimate existing mount.
+//
+// Alas, there is no good solution to that problem. A workaround, implemented
+// here, is to re-read mountinfo again if the mount is not found. With multiple
+// re-reads, the chances of repeatedly hitting the race are decreasing.
+//
+// Those re-reads are costly, so there is an additional optimization -- we
+// assume that if two subsequent reads returned the same number of entries,
+// there is no issue, and thus there is no need to re-read mountinfo again.
+// This is not a great heuristic but in case there are no parallel
+// mounts/unmounts it limits the number of reads to 2.
+//
+// Note that the main idea here is not to get two consecutive consistent
+// readings (with the same number of entries), but rather a complete reading.
+// Since there is no way to figure out if the reading is complete, we compare
+// the number of entries between the two readings.
+//
+// Finally, in case of massively parallel mounts/unmounts the number of entries
+// will always be different, and in this case maxTries is the main mechanism to
+// exit the re-reading loop.
+func mountedByMountinfoWithRetries(path string) (bool, error) {
+	var total, oldTotal, try int
+	const maxTries = 3
+
+again:
+	// The filter is the same as SingleEntryFilter,
+	// with the addition of counting total entries.
+	entries, err := GetMounts(func(m *Info) (bool, bool) {
+		if m.Mountpoint == path {
+			return false, true // don't skip, stop now
+		}
+		total++
+		return true, false // skip, keep going
+	})
+	if err != nil {
+		return false, err
+	}
+	if len(entries) > 0 {
+		return true, nil
+	}
```

Comment:
Would it not return a true without a retry if a mount point is found?

Reply:
Yes, the re-read is only for negative cases. Once the mount point is found, there's absolutely need for any kind of retries or re-reads (note it is different from kubernetes approach, which _always_ reads mountinfo at least twice).